provider/vsphere/environ_policy.go

```diff
@@ -23,8 +24,51 @@ func (env *environ) PrecheckInstance(ctx context.ProviderCallContext, args envir
 
 // PrecheckInstance is part of the environs.Environ interface.
 func (env *sessionEnviron) PrecheckInstance(ctx context.ProviderCallContext, args environs.PrecheckInstanceParams) error {
-	_, err := env.parsePlacement(ctx, args.Placement)
-	return err
+	if _, err := env.parsePlacement(ctx, args.Placement); err != nil {
```

Comment:
args.Placement may be "" now right?
So we should check for that?
And in other providers too?

Reply:
Well, we exit early above if both the constraints and the placement is empty. And parsePlacement handles empty placement correctly too (as checked by tests).

I'll go through the other providers to check they're not going to freak out being called with placement="".

Reply:
Checking through the implementations of PrecheckInstance under `provider` most of them are clearly written with the expectation that they'll be called even when there's no placement directive (other than maas that only looks at placement). They all have code paths that deal with placement being "".

Hmm, the manual provider's one always fails. So potentially this change would break that. I'm going to check that a bit more.

Reply:
I realised that the right thing to do here is not to do prechecking for machine placement because we're not creating instances.